File: backend/internal/service/notify.go
package service

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"gorm.io/gorm"
	"gorm.io/gorm/clause"

	"quant-system/backend/internal/model"
)

// NotifyService 飞书通知：
// - 定时推送统一由 quant-engine 通知调度器负责（本服务不参与定时）
// - 本服务仅处理用户主动触发的推送（测试卡片 / 手动执行交易卡片）
// - 配置读写 app_config 表（页面可改；与 Python 侧 load_config 同口径）
type NotifyService struct {
	db *gorm.DB
}

func NewNotifyService(db *gorm.DB) *NotifyService { return &NotifyService{db: db} }

// NotifyEventDTO 通知事件配置（GET 返回 / PUT 接收）
type NotifyEventDTO struct {
	EventKey     string  `json:"event_key"`
	Name         string  `json:"name"`
	Enabled      bool    `json:"enabled"`
	ScheduleType string  `json:"schedule_type"` // weekday / trading_day / event
	Weekdays     string  `json:"weekdays"`      // '1,2,3,4,5'（1=周一..7=周日）
	SendAt       *string `json:"send_at"`       // HH:MM；event 型为 null
	Template     string  `json:"template"`
}

// FeishuConfigDTO 飞书配置（app_config.feishu.*）
type FeishuConfigDTO struct {
	Enabled      bool   `json:"enabled"`
	WebhookURL   string `json:"webhook_url"`
	DashboardURL string `json:"dashboard_url"`
	Timeout      int    `json:"timeout"`
	MaxRetries   int    `json:"max_retries"`
	Secret       string `json:"secret"`  // 签名校验密钥；留空=不签名
	AtAll        bool   `json:"at_all"`  // 通知卡片 @所有人
}

// ---------- 通知事件配置 ----------

func (s *NotifyService) ListEvents() ([]NotifyEventDTO, error) {
	var rows []model.NotifyConfig
	if err := s.db.Order("event_key").Find(&rows).Error; err != nil {
		return nil, err
	}
	out := make([]NotifyEventDTO, 0, len(rows))
	for _, r := range rows {
		d := NotifyEventDTO{
			EventKey: r.EventKey, Name: r.Name, Enabled: r.Enabled,
			ScheduleType: r.ScheduleType, Weekdays: r.Weekdays, Template: r.Template,
		}
		if r.SendAt != nil {
			v := *r.SendAt
			if len(v) >= 5 {
				v = v[:5] // "19:30:00" → "19:30"
			}
			d.SendAt = &v
		}
		out = append(out, d)
	}
	return out, nil
}

// UpdateEvent 更新单个通知事件配置（event 型忽略 send_at/weekdays）
func (s *NotifyService) UpdateEvent(eventKey string, d NotifyEventDTO) error {
	if d.ScheduleType != "weekday" && d.ScheduleType != "trading_day" &&
		d.ScheduleType != "event" {
		return errors.New("schedule_type 仅支持 weekday / trading_day / event")
	}
	var row model.NotifyConfig
	if err := s.db.First(&row, "event_key = ?", eventKey).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return ErrNotifyEventMissing
		}
		return err
	}
	row.Enabled = d.Enabled
	row.ScheduleType = d.ScheduleType
	row.Template = d.Template
	row.Weekdays = ""
	row.SendAt = nil
	if d.ScheduleType == "weekday" {
		row.Weekdays = d.Weekdays
	}
	if d.ScheduleType == "weekday" || d.ScheduleType == "trading_day" {
		if d.SendAt == nil || *d.SendAt == "" {
			return errors.New("定时事件必须配置 send_at (HH:MM)")
		}
		t, err := time.Parse("15:04", *d.SendAt)
		if err != nil {
			return errors.New("send_at 格式应为 HH:MM")
		}
		s := t.Format("15:04:05")
		row.SendAt = &s
	}
	return s.db.Save(&row).Error
}

// ---------- 飞书配置 ----------

func (s *NotifyService) GetFeishuConfig() (FeishuConfigDTO, error) {
	var rows []model.AppConfig
	if err := s.db.Where("key LIKE 'feishu.%'").Find(&rows).Error; err != nil {
		return FeishuConfigDTO{}, err
	}
	vals := map[string]string{}
	for _, r := range rows {
		vals[r.Key] = r.Value
	}
	return FeishuConfigDTO{
		Enabled:      vals["feishu.enabled"] == "1" || vals["feishu.enabled"] == "true",
		WebhookURL:   vals["feishu.webhook_url"],
		DashboardURL: vals["feishu.dashboard_url"],
		Timeout:      atoiDef(vals["feishu.timeout"], 10),
		MaxRetries:   atoiDef(vals["feishu.max_retries"], 2),
		Secret:       vals["feishu.secret"],
		AtAll:        vals["feishu.at_all"] == "1" || vals["feishu.at_all"] == "true",
	}, nil
}

func (s *NotifyService) UpdateFeishuConfig(d FeishuConfigDTO) error {
	rows := []model.AppConfig{
		{Key: "feishu.enabled", Value: strconv.FormatBool(d.Enabled), ValueType: "bool"},
		{Key: "feishu.webhook_url", Value: d.WebhookURL, ValueType: "secret"},
		{Key: "feishu.dashboard_url", Value: d.DashboardURL, ValueType: "string"},
		{Key: "feishu.timeout", Value: strconv.Itoa(d.Timeout), ValueType: "int"},
		{Key: "feishu.max_retries", Value: strconv.Itoa(d.MaxRetries), ValueType: "int"},
		{Key: "feishu.secret", Value: d.Secret, ValueType: "secret"},
		{Key: "feishu.at_all", Value: strconv.FormatBool(d.AtAll), ValueType: "bool"},
	}
	return s.db.Clauses(clause.OnConflict{
		Columns:   []clause.Column{{Name: "key"}},
		DoUpdates: clause.AssignmentColumns([]string{"value", "value_type", "updated_at"}),
	}).Create(&rows).Error
}

// ---------- 发送 ----------

// Ready 是否可推送（启用且已配置 webhook）
func (s *NotifyService) Ready() (bool, error) {
	cfg, err := s.GetFeishuConfig()
	if err != nil {
		return false, err
	}
	return cfg.Enabled && cfg.WebhookURL != "", nil
}

// SendCard 发送飞书富文本卡片（同步；用户主动触发的推送）
func (s *NotifyService) SendCard(title, content, template, footer string) error {
	cfg, err := s.GetFeishuConfig()
	if err != nil {
		return err
	}
	if !cfg.Enabled || cfg.WebhookURL == "" {
		return errors.New("飞书通知未启用或未配置 webhook，请先在设置页配置")
	}
	if cfg.AtAll { // @所有人：lark_md 内容前置 @ 标签，触发全员提及通知
		content = `<at id="all"></at>` + "\n\n" + content
	}
	payload := buildCardPayload(title, content, template, footer)
	if cfg.Secret != "" { // 开启签名校验：请求体携带 timestamp + sign
		ts := time.Now().Unix()
		payload["timestamp"] = strconv.FormatInt(ts, 10)
		payload["sign"] = genSign(cfg.Secret, ts)
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	req, err := http.NewRequest("POST", cfg.WebhookURL, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	client := &http.Client{Timeout: time.Duration(cfg.Timeout) * time.Second}
	var lastErr error
	for attempt := 1; attempt <= cfg.MaxRetries; attempt++ {
		resp, err := client.Do(req)
		if err != nil {
			lastErr = err
			continue
		}
		var fb struct {
			Code int    `json:"code"`
			Msg  string `json:"msg"`
		}
		_ = json.NewDecoder(resp.Body).Decode(&fb)
		resp.Body.Close()
		if fb.Code == 0 {
			return nil
		}
		lastErr = fmt.Errorf("飞书返回错误 code=%d msg=%s", fb.Code, fb.Msg)
	}
	return fmt.Errorf("飞书推送失败（重试 %d 次）：%v", cfg.MaxRetries, lastErr)
}

// SendTest 发送测试卡片
func (s *NotifyService) SendTest() error {
	cfg, err := s.GetFeishuConfig()
	if err != nil {
		return err
	}
	content := "**测试卡片**\n\n如果能看到这条消息，说明飞书机器人配置正常。\n\n" +
		"[打开 Dashboard](" + dashboardURL(cfg.DashboardURL) + ")"
	return s.SendCard("Steady · 测试通知", content, "green", "测试推送")
}

// ---------- 辅助 ----------

var ErrNotifyEventMissing = errors.New("通知事件不存在")

func atoiDef(s string, def int) int {
	if n, err := strconv.Atoi(s); err == nil {
		return n
	}
	return def
}

func dashboardURL(d string) string {
	if d == "" {
		return "http://localhost"
	}
	return d
}

// genSign 飞书机器人签名：sign = base64(hmac_sha256(key=f"{ts}\n{secret}", data=空))
// 与 quant-engine/app/notify.py gen_sign 同算法（官方：string_to_sign 作 HMAC 密钥）。
func genSign(secret string, ts int64) string {
	mac := hmac.New(sha256.New, []byte(fmt.Sprintf("%d\n%s", ts, secret)))
	return base64.StdEncoding.EncodeToString(mac.Sum(nil))
}

func buildCardPayload(title, content, template, footer string) map[string]interface{} {
	elements := []interface{}{
		map[string]interface{}{
			"tag":  "div",
			"text": map[string]interface{}{"tag": "lark_md", "content": content},
		},
		map[string]interface{}{"tag": "hr"},
	}
	if footer != "" {
		elements = append(elements, map[string]interface{}{
			"tag":      "note",
			"elements": []interface{}{map[string]interface{}{"tag": "plain_text", "content": footer}},
		})
	}
	return map[string]interface{}{
		"msg_type": "interactive",
		"card": map[string]interface{}{
			"config":   map[string]interface{}{"wide_screen_mode": true},
			"header":   map[string]interface{}{"template": template, "title": map[string]interface{}{"tag": "plain_text", "content": title}},
			"elements": elements,
		},
	}
}

File: quant-engine/app/notify.py
"""飞书自定义机器人推送（统一通知传输层）

- 配置读 app_config 表（页面可改，值以库为准）；不读环境变量，未配置即禁用
- 异步发送（daemon 线程），失败不阻塞调用方；有限重试 + 超时
- webhook 日志脱敏，绝不打印完整 token
- 卡片结构：interactive 富文本，标题 + lark_md 摘要 + Dashboard 链接

配置键（app_config）：feishu.enabled / feishu.webhook_url / feishu.dashboard_url /
feishu.timeout / feishu.max_retries / feishu.secret / feishu.at_all（通知卡片 @所有人）
"""
import base64
import hashlib
import hmac
import json
import logging
import threading
import time
import urllib.request
from urllib.parse import urlsplit

from sqlalchemy import select

from app.models.tables import AppConfig

logger = logging.getLogger("notify")

# app_config 键 → 内部配置键
_KEYMAP = {
    "feishu.enabled": "enabled",
    "feishu.webhook_url": "webhook",
    "feishu.dashboard_url": "dashboard",
    "feishu.timeout": "timeout",
    "feishu.max_retries": "max_retries",
    "feishu.secret": "secret",
    "feishu.at_all": "at_all",
}


def gen_sign(secret: str, timestamp: int) -> str:
    """飞书机器人签名：sign = base64(hmac_sha256(key=f"{ts}\n{secret}", msg=b""))

    官方算法（机器人开启签名校验后必填，服务端用同样算法验签）：
    string_to_sign = "{timestamp}\n{secret}" 作为 HMAC 密钥，空消息体。
    """
    string_to_sign = f"{timestamp}\n{secret}"
    hmac_code = hmac.new(string_to_sign.encode("utf-8"),
                         digestmod=hashlib.sha256).digest()
    return base64.b64encode(hmac_code).decode("utf-8")


def load_config(db=None) -> dict:
    """加载飞书配置：仅从 app_config 表读取（页面可改，值以库为准）

    不读环境变量——按「业务配置全走页面，env 层只保留数据库凭据」原则；
    库未配置或读取失败 → enabled=False（通知不发送，不阻塞主业务）。
    """
    vals = {k: "" for k in _KEYMAP.values()}
    if db is not None:
        try:
            for row in db.execute(select(AppConfig)).scalars():
                internal = _KEYMAP.get(row.key)
                if internal and row.value not in (None, ""):
                    vals[internal] = row.value
        except Exception:
            db.rollback()
            logger.warning("读取 app_config 失败，按未配置处理", exc_info=True)
    return {
        "enabled": str(vals["enabled"]).strip().lower() in ("1", "true", "yes", "on"),
        "webhook": (vals["webhook"] or "").strip(),
        "dashboard": (vals["dashboard"] or "http://localhost").strip().rstrip("/"),
        "timeout": float(vals["timeout"] or 10),
        "max_retries": int(vals["max_retries"] or 2),
        "secret": (vals["secret"] or "").strip(),
        "at_all": str(vals["at_all"]).strip().lower() in ("1", "true", "yes", "on"),
    }


class FeishuNotifier:
    """飞书卡片推送器（一次性使用：传入配置，send_card 异步发送）"""

    def __init__(self, cfg: dict):
        self.enabled = bool(cfg.get("enabled"))
        self.webhook = str(cfg.get("webhook") or "").strip()
        self.dashboard = str(cfg.get("dashboard") or "http://localhost").rstrip("/")
        self.timeout = float(cfg.get("timeout") or 10)
        self.max_retries = int(cfg.get("max_retries") or 2)
        self.secret = str(cfg.get("secret") or "").strip()
        self.at_all = bool(cfg.get("at_all"))

    def url(self, path: str = "/") -> str:
        return self.dashboard + (path if path.startswith("/") else "/" + path)

    @property
    def ready(self) -> bool:
        """可发送：启用 且 已配置 webhook"""
        return self.enabled and bool(self.webhook)

    def mask(self) -> str:
        try:
            parts = urlsplit(self.webhook)
            return f"{parts.scheme}://{parts.netloc}{parts.path}?****"
        except Exception:
            return "****"

    def send_card(self, title: str, content_md: str, template: str = "blue",
                  footer: str | None = None, wait: bool = False) -> bool:
        """发送富文本卡片。默认异步（不阻塞主流程）；wait=True 同步并返回是否送达。"""
        if not self.ready:
            logger.debug("飞书通知未启用或未配置 webhook，跳过推送")
            return False
        payload = self._build_card(title, content_md, template, footer)
        if self.secret:  # 开启签名校验：请求体携带 timestamp + sign
            ts = int(time.time())
            payload["timestamp"] = str(ts)
            payload["sign"] = gen_sign(self.secret, ts)
        if wait:
            return self._post(payload)
        threading.Thread(target=self._post, args=(payload,), daemon=True).start()
        return True

    def send_test(self, wait: bool = True) -> bool:
        content = (
            "**测试卡片**\n\n如果能看到这条消息，说明飞书机器人配置正常。\n\n"
            f"[打开 Dashboard]({self.url('/')})"
        )
        return self.send_card("Steady · 测试通知", content,
                              template="green", footer="测试推送", wait=wait)

    def _build_card(self, title: str, content_md: str, template: str,
                    footer: str | None) -> dict:
        if self.at_all:  # @所有人：lark_md 内容前置 @ 标签，触发全员提及通知
            content_md = '<at id="all"></at>\n\n' + content_md
        elements = [
            {"tag": "div", "text": {"tag": "lark_md", "content": content_md}},
            {"tag": "hr"},
        ]
        if footer:
            elements.append({"tag": "note",
                             "elements": [{"tag": "plain_text", "content": footer}]})
        return {
            "msg_type": "interactive",
            "card": {
                "config": {"wide_screen_mode": True},
                "header": {"template": template,
                           "title": {"tag": "plain_text", "content": title}},
                "elements": elements,
            },
        }

    def _post(self, payload: dict) -> bool:
        data = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            self.webhook, data=data,
            headers={"Content-Type": "application/json"})
        for attempt in range(1, self.max_retries + 1):
            try:
                with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                    body = json.loads(resp.read().decode("utf-8") or "{}")
                if body.get("code") == 0:
                    logger.info("飞书推送成功（%s）", self.mask())
                    return True
                logger.warning("飞书推送失败 第%d/%d次: HTTP %s code=%s msg=%s",
                               attempt, self.max_retries, resp.status,
                               body.get("code"), body.get("msg", ""))
            except Exception as e:  # 超时/连接错误等：任何异常都不上抛
                logger.warning("飞书推送异常 第%d/%d次: %s", attempt,
                               self.max_retries, e)
            if attempt < self.max_retries:
                time.sleep(attempt)
        logger.error("飞书推送失败，已达最大重试次数（%s）", self.mask())
        return False
